include/xtensor/xsort.hpp

```diff
@@ -144,6 +144,148 @@ namespace xt
+                std::sort(inds.data() + inds_offset, inds.data() + inds_offset + inds_secondary_stride, comp);
+            }
+        }
+    }
+
+    template <class E>
+    auto argsort(const xexpression<E>& e, placeholders::xtuph /*t*/)
+    {
+        using value_type = typename E::value_type;
+        using result_type = typename detail::argsort_result_type<E>::type;
+
+        const auto de = e.derived_cast();
+
+        result_type result;
+        result.resize({de.size()});
+        auto comp = [&de](std::size_t x, std::size_t y) {
+            return de[x] < de[y];
+        };
+        std::iota(result.begin(), result.end(), 0);
+        std::sort(result.begin(), result.end(), comp);
+
+        return result;
+    }
+
+    /**
+     * Argsort xexpression (optionally along axis)
+     * Performs an indirect sort along the given axis. Returns an xarray
+     * of indices of the same shape as e that index data along the given axis in
+     * sorted order.
+     *
+     * @param e xexpression to argsort
+     * @param axis axis along which argsort is performed
+     *
+     * @return argsorted index array
+     */
+    template <class E>
+    auto argsort(const xexpression<E>& e, std::size_t axis)
+    {
+        using eval_type = typename E::temporary_type;
+        using value_type = typename E::value_type;
+        using result_type = typename detail::argsort_result_type<E>::type;
+
+        const auto& de = e.derived_cast();
+
+        if (de.dimension() == 1)
+        {
+            return argsort(de, xnone());
+        }
+
+        eval_type ev;
+
+        if (axis != detail::leading_axis(ev))
+        {
+            auto axis_numbers = arange<std::size_t>(de.shape().size());
+            std::vector<std::size_t> permutation(axis_numbers.begin(), axis_numbers.end());
+            permutation.erase(permutation.begin() + std::ptrdiff_t(axis));
+            if (de.layout() == layout_type::row_major)
+            {
+                permutation.push_back(axis);
+            }
+            else
+            {
+                permutation.insert(permutation.begin(), axis);
+            }
+
+            // TODO find a more clever way to get reverse permutation?
+            std::vector<std::size_t> reverse_permutation;
+            for (auto el : axis_numbers)
+            {
+                auto it = std::find(permutation.begin(), permutation.end(), el);
+                reverse_permutation.push_back(std::size_t(std::distance(permutation.begin(), it)));
+            }
+
+            ev = transpose(de, permutation);
+            result_type res(ev.shape(), ev.layout());
+            detail::argsort_over_leading_axis(ev, res);
+            ev = transpose(ev, reverse_permutation);
+            res = transpose(res, reverse_permutation);
+            return res;
+        }
+        else
+        {
+            ev = de;
+            result_type res(de.shape(), de.layout());
+            detail::argsort_over_leading_axis(ev, res);
```

Comment:
with argsort_over_leading_axis taking its first parameter as a const ref, no need to copy de in ev anymore, you can pass de directly.